### DISCUSSION TITLE
Why am I getting a linker error with the cpp-base64 port?

### DISCUSSION THREAD

**Discussion by BullyWiiPlaza** [microsoft/vcpkg#27539] [General]
When I try to use the [`cpp-base64`](https://github.com/ReneNyffenegger/cpp-base64) port with `CMake` and Visual Studio, I run into a linker error: 

    LNK2019	unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl base64_decode(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,bool)" (?base64_decode@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBV12@_N@Z) referenced in function

The suggested `CMake` entries to add by `vcpkg` are as follows, which I also added to my `CMakeLists.txt`:
```
find_path(CPP_BASE64_INCLUDE_DIRS "cpp-base64/base64.cpp")
target_include_directories(${PROJECT_NAME} PRIVATE ${CPP_BASE64_INCLUDE_DIRS})
```
The library is very simple and only consists of a `base64.cpp` and `base64.h` file.

Using `CMake`, how can I link against this library successfully? I assume this is a very simple fix but why is it not as straight-forward and painless like with other libraries where all you basically need to do is install the package, copy/paste the `CMake` entries into your `CMakeLists.txt` and then it compiles/links?

Related StackOverflow question: https://stackoverflow.com/questions/61119878

**Comment by BullyWiiPlaza** [comment]
Bump

**Reply by JavierMatosD** [reply]
Hi @BullyWiiPlaza, would you mind including your CMakeLists.txt?

**Reply by BullyWiiPlaza** [reply]
@JavierMatosD My `CMakeLists.txt` file does not contain anything regarding the `cpp-base64` port at all, I just ended up going with `#include <cpp-base64/base64.cpp>` in one `.cpp` file so far.

**Comment by Osyotr** [comment]
This port is cursed.
It should be static-only on windows, because it does not export symbols.
Users will likely need to build it with C++17 flags to be able to use `string_view` overloads. (=> use custom triplet)

Alternative is to include `cpp-base64/base64.cpp` in exactly one translation unit in your project.

**Reply by BullyWiiPlaza** [reply]
If you think you can improve the port, feel free. It would be great to just write `#include <cpp-base64/base64.hpp>` in your code and it links successfully like with many other libraries.

**Reply by dg0yt** [reply]
> It would be great to just write `#include <cpp-base64/base64.hpp>` in your code and it links successfully like with many other libraries.

This is possible for pure header-only ports. But cpp-base64 seems to come with code that must be compiled once. The only reliable way for vcpkg to control this "once" is to provide a library.
"`#include <cpp-base64/base64.cpp>` in one .cpp file" is an alternative to a library if you are sure that it is never done in any dependencies.
